databricks/koalas/frame.py

```diff
@@ -9659,6 +9657,117 @@ class  max_speed
+        injection if you pass user input to this function.
+
+        Parameters
+        ----------
+        expr : str
+            The expression string to evaluate.
+        inplace : bool, default False
+            If the expression contains an assignment, whether to perform the
+            operation inplace and mutate the existing DataFrame. Otherwise,
+            a new DataFrame is returned.
+
+        Returns
+        -------
+        The result of the evaluation.
+
+        See Also
+        --------
+        DataFrame.query : Evaluates a boolean expression to query the columns
+            of a frame.
+        DataFrame.assign : Can evaluate an expression or function to create new
+            values for a column.
+        eval : Evaluate a Python expression as a string using various
+            backends.
+
+        Examples
+        --------
+        >>> df = ks.DataFrame({'A': range(1, 6), 'B': range(10, 0, -2)})
+        >>> df
+           A   B
+        0  1  10
+        1  2   8
+        2  3   6
+        3  4   4
+        4  5   2
+        >>> df.eval('A + B')
+        0    11
+        1    10
+        2     9
+        3     8
+        4     7
+        Name: 0, dtype: int64
+
+        Assignment is allowed though by default the original DataFrame is not
+        modified.
+
+        >>> df.eval('C = A + B')
+           A   B   C
+        0  1  10  11
+        1  2   8  10
+        2  3   6   9
+        3  4   4   8
+        4  5   2   7
+        >>> df
+           A   B
+        0  1  10
+        1  2   8
+        2  3   6
+        3  4   4
+        4  5   2
+
+        Use ``inplace=True`` to modify the original DataFrame.
+
+        >>> df.eval('C = A + B', inplace=True)
+        >>> df
+           A   B   C
+        0  1  10  11
+        1  2   8  10
+        2  3   6   9
+        3  4   4   8
+        4  5   2   7
+        """
+        if isinstance(self.columns, pd.MultiIndex):
+            raise ValueError("`eval` is not supported for multi-index columns")
+        inplace = validate_bool_kwarg(inplace, "inplace")
+        is_series = False
+        is_scalar_ = False
+
+        # Since `eva_func` doesn't have a type hint, inferring the schema is always preformed
+        # in the `map_in_pandas`. Hence, the variables `is_seires` and `is_scalar_` can be updated.
+        def eval_func(pdf):
+            nonlocal is_series
+            nonlocal is_scalar_
+            result_inner = pdf.eval(expr, inplace=inplace)
+            if inplace:
+                result_inner = pdf
+            if isinstance(result_inner, pd.Series):
+                is_series = True
+                result_inner = result_inner.to_frame()
+            elif is_scalar(result_inner):
+                is_scalar_ = True
+                result_inner = pd.Series(result_inner).to_frame()
+            return result_inner
+
+        result = self.map_in_pandas(eval_func)
+        if inplace:
```

Comment:
How about:

```python
if inplace:
    # Here, it always returns a DataFrame because the error is thrown during schema inference
    # from pandas.
    self._internal = result._internal
elif should_return_series:
    series_name = result._internal.column_labels[0][0]
    return result[series_name]
elif should_return_scalar:
    series_name = result._internal.column_labels[0][0]
    return result[series_name][0]
else:
    return result
```

Reply:
Looks clearer. thanks for the comment 👍